jwst/msaflagopen/msaflag_open.py

```diff
@@ -0,0 +1,235 @@
+        science data with DQ flags of affected modified
+
+    """
+    #
+    # Use the machinery in assign_wcs to create a WCS object for the bad shutters
+    pipeline = slitlets_wcs(input_datamodel, wcs_refnames, failed_slitlets)
+    wcs = WCS(pipeline)
+    # Create output as a copy of the input science data model so we can overwrite
+    # the wcs with the wcs for the failed open shutters
+    temporary_copy = input_datamodel.copy()
+    temporary_copy.meta.wcs = wcs
+    dq_array = input_datamodel.dq
+    for slitlet in failed_slitlets:
+        #
+        # Pick the WCS for this slitlet from the WCS of the exposure
+        thiswcs = nrs_wcs_set_input(temporary_copy, slitlet.name)
+        #
+        # Convert the bounding box for this slitlet to a set of indices to use as a slice
+        xmin, xmax, ymin, ymax = boundingbox_to_indices(temporary_copy,
+                                                        thiswcs.bounding_box)
+        #
+        # Make a grid of points within the slice
+        y_indices, x_indices = np.mgrid[ymin:ymax, xmin:xmax]
+        #
+        # Calculate the arrays of coordinates for each pixel in the slice
+        coordinate_array = thiswcs(x_indices, y_indices)
+        #
+        # The coordinate_array is a tuple of arrays, one for each output coordinate
+        # In this case there should be 3 arrays, one each for RA, Dec and Wavelength
+        # For pixels outside the slitlet, the arrays have NaN
+        #
+        # Make a subarray from these coordinate arrays by setting pixels that aren't
+        # NaN to FAILDOPENFLAG, the rest to 0
+        dq_subarray = wcs_to_dq(coordinate_array, FAILEDOPENFLAG)
+        #
+        # bitwise-or this subarray with the slice in the original exposure's DQ array
+        dq_array = or_subarray_with_array(dq_array, dq_subarray, xmin, xmax, ymin, ymax)
+    #
+    # Set the dq array of the input datamodel to the corrected dq array
+    input_datamodel.dq = dq_array
+    return input_datamodel
+
+def boundingbox_to_indices(data_model, bounding_box):
+    """
+    Takes a bounding_box (tuple of tuples: ((x1, x2), (y1, y2)) and
+    a datamodel and calculates the range of indices in the X and Y dimensions
+    of the overlap between the bounding box and the datamodel's data array.
+
+    Parameters
+    ----------
+    data_model: data model object
+        the input science datamodel
+
+    bounding_box: Tuple of tuples
+        Bounding box returned from wcs object
+
+    Returns
+    -------
+    xmin, xmax, ymin, ymax: integers
+        Range of indices of overlap between science data array and bounding box
+
+    """
+    nrows, ncols = data_model.data.shape
+    ((x1, x2), (y1, y2)) = bounding_box
+    xmin = int(min(x1, x2))
+    xmin = max(xmin, 0)
+    xmax = int(max(x1, x2)) + 1
+    xmax = min(xmax, ncols)
+    ymin = int(min(y1, y2))
+    ymin = max(ymin, 0)
+    ymax = int(max(y1, y2)) + 1
+    ymax = min(ymax, nrows)
+    return (xmin, xmax, ymin, ymax)
```

Comment:
Yes, before I got the assign_wcs.nirspec code working properly some of the bounding boxes returned extended outside the data array.  If there's code in assign_wcs that does the same thing it would appear to be redundant.  And I wasn't aware of the grid_from_bounding_box method until I noticed it very recently - is it new?